### PR TITLE
added unit tests using testthat

### LINKED ISSUE #70: Add unit tests
Add unit tests

### PR DESCRIPTION
**Pull request by mona-n**
fixes #70

### CODE REVIEW

File: R/GRN.R [hunk 1]
```diff
@@ -15,7 +15,8 @@
 #' data('DEGsmatrix')
 #' data('dataFilt')
 #' dataDEGs <- DEGsmatrix
-#' dataGRN <- GRN(TFs = rownames(dataDEGs)[1:10],
+#' set.seed(317)
```
Comment (mtiberti) [inline R/GRN.R]: I don't think we should be committing setting the seed here - this part of code is just for the examples, not tests

File: tests/testthat/test-DMA.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+               dataPRA = dataPRA,
+               runCscape = FALSE,
+               results_folder = ".")
+
+# Test that output of DMA is as expected
+test_that("output of DMA is a list with 2 elements", {
+  expect_type(dataDMA, "list")
+  expect_named(dataDMA, c("TSG", "OCG"))
+  expect_true(all(sapply(dataDMA, is.character)))
+})
```
Comment (mtiberti) [inline tests/testthat/test-DMA.R:22]: should this not also be changed to check the results rather than the format?

File: tests/testthat/test-GLS.R [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Test GLS function
```
Comment (mtiberti) [inline tests/testthat/test-GLS.R:1]: this test is potentially problematic in the long term - I'm sure it works _now_ but it might fail in the future just because new papers on these genes get released.

What if (just for this test) we limit the search date to sometimes in the past? I checked on PubMed and you can e.g. use these keywords:

```
("1980/01/01"[Date - Publication] : "2023/01/01"[Date - Publication])
```

this should help keep the result mostly consistent

File: tests/testthat/test-GRN.R [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+})
+
+# Test that names of elements in GRN are correct
+test_that("names of elements in GRN output are correct", {
+  expect_equal(names(dataGRN_test), c("miTFGenes", "maxmi"))
+})
+
+# Test that output of GRN is as expected compared to reference / example data
+test_that("GRN output is identical to reference point", {
+  expect_equal(dataGRN_test, dataGRN, tolerance = 0.1)
```
Comment (mtiberti) [inline tests/testthat/test-GRN.R]: 0.1 is a bit high as tolerance for MI; ideally we should define tolerance depending on the significant digits of the output data type. Can we do at least third decimal?

File: tests/testthat/test-GRN.R [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+})
+
+
+
+
+
+
+
+
```
Comment (mtiberti) [inline tests/testthat/test-GRN.R]: bunch of newlines to remove

File: tests/testthat/test-PRA.R [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (mtiberti) [inline tests/testthat/test-PRA.R]: bunch of newlines to remove

File: tests/testthat/test-URA.R [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (mtiberti) [inline tests/testthat/test-URA.R]: bunch of newlines to remove

File: NEWS.md [hunk 8]
```diff
@@ -1,3 +1,9 @@
+# Moonlight2R 0.99.3
+
+## Summary
+
+* added tests with testthat
```
Comment (mtiberti) [inline NEWS.md:5]: I see you also changed some of the example data/available data, can you add a note here?

File: tests/testthat/test-DMA.R [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (mtiberti) [inline tests/testthat/test-DMA.R]: extra newlines to be removed

File: tests/testthat/test-FEA.R [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  expect_equal(dataFEA_test, dataFEA)
+})
+
+
+
+
+
+
+
+
```
Comment (mtiberti) [inline tests/testthat/test-FEA.R]: extra newlines to be removed

**Review by mtiberti — CHANGES_REQUESTED** [review]

File: R/GRN.R (hunk 1 above)
Comment (mona-n) [reply R/GRN.R]: I added the seed here so we know what seed generated the example data and so that we would know how to replicate the example data. Should I place it somewhere else?

File: tests/testthat/test-DMA.R (hunk 2 above)
Comment (mona-n) [reply tests/testthat/test-DMA.R:22]: Ideally yes, I just couldn't figure out how to do this due to the fine balance between having as small as possible example data while still getting meaningful results of the example data. So right now the example data of DMA contains one gene. When I regenerate the example data, the DMA data will be an empty list because the example MAF file and other example data is so small. I couldn't regenerate non-empty DMA data while still keeping the example data files rather small (I tried a lot of different things to generate non-empty DMA example data). 

If I were to add checking the results in this test like in the other tests, it would always fail because the output of the test is empty whereas the example data DMA is non-empty. I don't think it makes sense to have empty DMA example data

Let me know if you have any ideas on how to solve this

File: tests/testthat/test-GRN.R (hunk 4 above)
Comment (mona-n) [reply tests/testthat/test-GRN.R]: It doesn't seem to be possible because it can lead to failure of the test because sometimes the MI can be e.g. 0.2998 and then the next time it can be 0.300, hence the tolerance = 0.1. I believe it is the stochastic nature of the parmigene package that is at play. At least by setting the seed like this we are getting consistent results with tolerance = 0.1. 
In R, when you use the seed you need to set it right before running a line.
If you have other ideas let me know

File: tests/testthat/test-GLS.R (hunk 3 above)
Comment (mona-n) [reply tests/testthat/test-GLS.R:1]: updated

File: tests/testthat/test-GRN.R (hunk 5 above)
Comment (mona-n) [reply tests/testthat/test-GRN.R]: removed

File: tests/testthat/test-PRA.R (hunk 6 above)
Comment (mona-n) [reply tests/testthat/test-PRA.R]: removed

File: tests/testthat/test-URA.R (hunk 7 above)
Comment (mona-n) [reply tests/testthat/test-URA.R]: removed

File: NEWS.md (hunk 8 above)
Comment (mona-n) [reply NEWS.md:5]: added

File: tests/testthat/test-DMA.R (hunk 9 above)
Comment (mona-n) [reply tests/testthat/test-DMA.R]: removed

File: tests/testthat/test-FEA.R (hunk 10 above)
Comment (mona-n) [reply tests/testthat/test-FEA.R]: removed

File: tests/testthat/test-GRN.R (hunk 4 above)
Comment (mtiberti) [reply tests/testthat/test-GRN.R]: I don't think that's how `expect_equal` works, it checks if the difference is at least of 0.1, e.g.

```
> expect_equal(0.2998, 0.3, tolerance=0.1)
>

> expect_equal(0.2998, 0.3999, tolerance=0.1)
Error: 0.2998 not equal to 0.3999.
1/1 mismatches
[1] 0.3 - 0.4 == -0.1
```

if the difference between two runs is as large as 0.1, I think our test is not going to be very accurate. What we would need to do is to diminish the uncertainty of the result here rather than have very large (relatively speaking) allowed tolerance. I tried having a look at the source code of `knnmi.cross`, it seems to me that it is non-deterministic in generating a random seed that is used to generate noise. The order of magnitude of noise can be set in the `knnmi.cross` function call, I did a few tests and I think it becomes completely reproducible if we set it to `0`, e.g.:

```
library(parmigene)
library(testthat)

m1 = random_matrix <- matrix(runif(n=10000, min=1, max=20), nrow=1000, ncol=100)
m2 = random_matrix <- matrix(runif(n=10000, min=1, max=20), nrow=1000, ncol=100)

d1 = knnmi.cross(m1, m2, k=2, noise=0)

d2 = knnmi.cross(m1, m2, k=2, noise=0)

expect_equal(d1, d2)
```

I've run this several times and always passes for me, and returns the same result in independent runs. So we should try setting random seed *and* setting noise, because the output of `GRN` also depends on the random seed for the non-parmigene part. Worth a try?

Comment (mtiberti) [reply tests/testthat/test-GRN.R]: we could e.g. add a new small dataset with these "reproducible" results specifically for this test, not sure it is worth resaving it as the new GRN dataset and then redoing all the other results/tests in the pipeline, if they are reproducible

File: tests/testthat/test-DMA.R (hunk 2 above)
Comment (mtiberti) [reply tests/testthat/test-DMA.R:22]: let's leave it as is for now, we can see how our reviewer reacts to this

**Review by mtiberti — APPROVED** [review]
